lib/multiplexer/multiplexer.go

```diff
@@ -246,44 +256,145 @@ func (m *Mux) detectAndForward(conn net.Conn) {
 	listener.HandleConnection(m.context, connWrapper)
 }
 
-func detect(conn net.Conn, enableProxyProtocol bool) (*Conn, error) {
+type PROXYSigner interface {
+	SignPROXY(p jwt.PROXYSignParams) (string, error)
+}
+
+func getTCPAddr(a net.Addr) net.TCPAddr {
+	addr, ok := a.(*net.TCPAddr)
+	if ok { // Hot path
+		return *addr
+	}
+
+	parsedAddr := utils.FromAddr(a)
+	return net.TCPAddr{
+		IP:   net.ParseIP(parsedAddr.Host()),
+		Port: parsedAddr.Port(-1),
+	}
+}
+
+func GetSignedPROXYHeader(sourceAddress, destinationAddress net.Addr, clusterName string, signingCert []byte, signer PROXYSigner) ([]byte, error) {
+	sAddr := getTCPAddr(sourceAddress)
+	dAddr := getTCPAddr(destinationAddress)
+	if sAddr.Port < 0 || dAddr.Port < 0 {
```

Comment:
If `getTCPAddr` can be fallible, make it return an error, IMO.

Reply:
I feel like code is a bit more compact and even clearer like this, would prefer to leave it like this, if it's ok with you.